### DISCUSSION TITLE
Curious behaviour - using a lazy='joined' relationship in a hybrid_property works only when querying the entire object

### DISCUSSION THREAD

**Discussion by fraser-langton** [sqlalchemy/sqlalchemy#8855] [Usage Questions] [answered]
I am curious to why the below `weird_query` does not yield a query with the parent relationship loaded (I want it to come out like `working_query`)

My gut feeling is that relationships are only being loaded when the "entire" object is being queried, ie `session.query(Child)` whereas when `session.query(Child.parent_val)` relationships won't be loaded as we are only loading a single value not the ORM object itself?

If this is a case, is there a way I can force a hybrid expression when called as a lone attribute to join with a related table?
I am reading up on using a [Transformer](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#building-transformers) - I will see how that goes... 

FYI I want to force this at the model level, ie I have many existing queries that I would like to keep backwards compatible. I know this is solved simply in the below example by joining. 

### Working Example
```
import sqlalchemy as sa

from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import relationship

from sqlalchemy.orm import Query


Base = declarative_base()


class Parent(Base):
    __tablename__ = "parent_table"
    id = sa.Column(sa.Integer, primary_key=True)
    val = sa.Column(sa.Text)

    children = relationship("Child", back_populates="parent")


class Child(Base):
    __tablename__ = "child_table"
    id = sa.Column(sa.Integer, primary_key=True)
    parent_id = sa.Column(sa.Integer, sa.ForeignKey("parent_table.id"))
    val = sa.Column(sa.Text)

    parent = relationship("Parent", back_populates="children", lazy='joined')

    @hybrid_property
    def parent_val(self):
        return self.parent.val

    @parent_val.setter
    def parent_val(self, val):
        self.parent.val = val

    @parent_val.expression
    def parent_val(self):
        return Parent.val


working_query = Query([Child])
"""
SELECT child_table.id, child_table.parent_id, child_table.val, parent_table_1.id AS id_1, parent_table_1.val AS val_1 
FROM child_table
LEFT OUTER JOIN parent_table AS parent_table_1
	ON parent_table_1.id = child_table.parent_id;
"""

weird_query = Query([Child.id, Child.parent_id, Child.val, Child.parent_val])
"""
SELECT child_table.id, child_table.parent_id, child_table.val, parent_table.val AS val_1 
FROM child_table, parent_table;
"""
```

**Comment by zzzeek** [comment, marked as answer]
> I am curious to why the below `weird_query` does not yield a query with the parent relationship loaded (I want it to come out like `working_query`)
> 
> My gut feeling is that relationships are only being loaded when the "entire" object is being queried, ie `session.query(Child)` whereas when `session.query(Child.parent_val)` relationships won't be loaded as we are only loading a single value not the ORM object itself?

that's correct, the joinedload() eager loader is used only to affect how the production of newly loaded `Child` objects will proceed.  if you aren't loading `Child` objects then loader options are irrelevant.

> 
> If this is a case, is there a way I can force a hybrid expression when called as a lone attribute to join with a related table? I am reading up on using a [Transformer](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#building-transformers) - I will see how that goes...

There's no way that the presence of a hybrid can automatically cause a particular kind of JOIN to be added to the query.   Typically you need to be explicit as to what SQL you want to emit and what FROM clause you'd like:

```py
q = Query([Child.id, Child.parent_id, Child.val, Child.parent_val]).join(Child.parent)
```

to do it all "within the hybrid" would mean the hybrid has to emit a column expression that is against a correlated scalar subquery, which is not typically as efficient from a SQL perspective.  


> 
> FYI I want to force this at the model level, ie I have many existing queries that I would like to keep backwards compatible. I know this is solved simply in the below example by joining.

JOINs are very particular so SQLAlchemy steers clear of helpers that try to guess where a JOIN should be added.      it's conceivable that one could write some event handlers that try to add JOINs like this when certain entities are detected in the columns clause, but heuristics like this are very difficult to come up with and maintain and ultimately make it pretty likely that things will misbehave someday.

> #### My goal is for the `parent_val` to behave as though it's really a column on Child (using the model you wouldn't know any different)

yeah, this is a common thing, I think however you'd need to establish a routine that adds the column expression and the JOIN based on a common function of some kind.

**Reply by fraser-langton** [reply]
@zzzeek Thanks for the response!
